# Practicum
# Test task for a code reviewer in Data Science
### Jerome Smith, 25/02/2023

Each question of the test is answered in its own Jupyter notebook cell below.

## Task 1. Working with data

In [192]:
# Libraries and settings
import pandas as pd
import statistics as st
from pandasql import sqldf
import datetime as dt
import warnings
warnings.filterwarnings('ignore')

In [193]:
# Q 1.1 Download the data set movie_metadata.csv, which contains data about films from IMDb (Internet Movie Database).
# Load movie_metadata.csv into data frame and inspect.
df_movie = pd.read_csv('movie_metadata.csv')
df_movie.dtypes
print(df_movie.shape[0], 'rows')
df_movie.head()

5043 rows


color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating        budget  \
0               3054.0  English      USA           PG-13  237000000.0$   
1               1238.0  English      USA           PG-13  300000000.0$   
2                994.0  English       UK           PG-13  245000000.0$   
3               2701.0  English      USA           PG-13  250000000.0$   
4                  NaN      NaN      NaN             NaN          0.0$   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [194]:
# Q 1.2 The duration column contains data on the film length. How many missing values are there in this column?
na = df_movie[pd.isna(df_movie['duration'])].shape[0]
print('Q 1.2: There are', na, 'missing values in the duration column.')

Q 1.2: There are 15 missing values in the duration column.


In [195]:
# Q 1.3 Replace the missing values in the duration column with the median value for this column.
median = df_movie['duration'].median()
df_movie.loc[pd.isna(df_movie['duration']), 'duration'] = median
df_movie.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN     103.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating        budget  \
0               3054.0  English      USA           PG-13  237000000.0$   
1               1238.0  English      USA           PG-13  300000000.0$   
2                994.0  English       UK           PG-13  245000000.0$   
3               2701.0  English      USA           PG-13  250000000.0$   
4                  NaN      NaN      NaN             NaN          0.0$   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [196]:
# Q 1.4 What is the average film length? Give the answer as a floating-point figure rounded to two decimal places.
mean = round(df_movie['duration'].mean(), 2)
print('Assuming that the mean must be calculated AFTER substituting missing values with the median, the mean duration is', mean, 'minutes.')

Assuming that the mean must be calculated AFTER substituting missing values with the median, the mean duration is 107.19 minutes.


In [197]:
# Q 1.5 Create a movie_duration_category column, which will contain three categories depending on the film length:
# Category "1. <90" if the film is less than 90 minutes long
# Category "2. 90–120" if the film is between 90 minutes and two hours long (inclusively)
# Category "3. >120" if the film is more than two hours long
df_movie['movie_duration_category'] = '1. <90'
df_movie.loc[df_movie['duration'] >= 90, 'movie_duration_category'] = '2. 90–120'
df_movie.loc[df_movie['duration'] > 120, 'movie_duration_category'] = '3. >120'
df_movie[['duration', 'movie_duration_category']]

duration movie_duration_category
0        178.0                 3. >120
1        169.0                 3. >120
2        148.0                 3. >120
3        164.0                 3. >120
4        103.0               2. 90–120
...        ...                     ...
5038      87.0                  1. <90
5039      43.0                  1. <90
5040      76.0                  1. <90
5041     100.0               2. 90–120
5042      90.0               2. 90–120

[5043 rows x 2 columns]

In [198]:
# Q 1.6 Build a summary table for films released after 2000 (inclusively), to list the numbers of films:
# Table rows: year
# Table columns: movie duration category ("<90", "90–120", ">120")
# The year of release should be displayed in the YYYY format.
df = df_movie.loc[df_movie['title_year'] >= 2000 , ['title_year', 'movie_duration_category']]
df['number_of_films'] = 1
df['title_year'] = df['title_year'].astype('int')
df_summary = pd.pivot_table(  data = df
                            , index = ['title_year']
                            , columns = ['movie_duration_category']
                            , values = ['number_of_films']
                            , aggfunc = 'sum'
                           )
df_summary.columns = ["<90", "90–120", ">120"]
df_summary

<90  90–120  >120
title_year                   
2000         25     112    34
2001         29     120    39
2002         36     146    27
2003         31     108    30
2004         30     142    42
2005         31     142    48
2006         40     146    53
2007         31     130    43
2008         29     160    36
2009         42     178    40
2010         26     171    33
2011         36     150    39
2012         41     139    41
2013         38     145    54
2014         36     171    45
2015         34     145    47
2016          8      79    19

Q 1.7 How many films between 90 minutes and two hours long were released in 2008?

Answer: By inspecting the summary table of question 1.6, it can be seen that the number of films between 90 minutes and two hours long released in 2008 is 160.

In [199]:
# Q 1.8 The plot_keywords column holds keywords characterizing the film's plot. Using the data in this column, create a column called movie_plot_category, to contain four categories depending on the key words in the column:
# Category "love_and_death" if the keywords include both "love" and "death"
# Category "love" if the keywords include the word "love"
# Category "death" if the keywords include the word "death"
# Category "other" if the keywords do not meet the conditions above

df_movie['movie_plot_category'] = 'other'
df_movie.loc[df_movie.plot_keywords.str.find('love') > -1, 'movie_plot_category'] = 'love'
df_movie.loc[df_movie.plot_keywords.str.find('death') > -1, 'movie_plot_category'] = 'death'
df_movie.loc[(df_movie.plot_keywords.str.find('love') > -1) & (df_movie.plot_keywords.str.find('death') > -1), 'movie_plot_category'] = 'love_and_death'
pd.set_option('display.max_rows', 20)
print(df_movie.loc[df_movie.movie_plot_category != 'other', ['plot_keywords', 'movie_plot_category']])

                                          plot_keywords movie_plot_category
9                      blood|book|love|potion|professor                love
26                         artist|love|ship|titanic|wet                love
31        death|doctor|scientist|super villain|tentacle               death
39    costumed hero|death of girlfriend|masked vigil...               death
44               death row|future|machine|rescue|skynet               death
...                                                 ...                 ...
5011     business trip|love|misogynist|office|secretary                love
5014     death|first part|killing spree|massacre|murder               death
5020               avatar|college|death|tron|university               death
5024    arch villain|game of death|kidnapping|superhero               death
5035                 assassin|death|guitar|gun|mariachi               death

[422 rows x 2 columns]


In [200]:
# Q 1.9 The imdb_score column shows a viewer rating for the film. Build a table to reflect the average rating of films depending on which movie_plot_category category they belong to.
sqldf('select movie_plot_category, avg(imdb_score) as avg_rating from df_movie group by movie_plot_category')

movie_plot_category  avg_rating
0               death    6.535465
1                love    6.580769
2      love_and_death    6.506250
3               other    6.431422

Q 1.10 What is the average rating of films in the "love" category? Give the answer as a floating point figure rounded to two decimal places.

Answer: By inspection of the result set of question 1.9, the average rating of films in the "love" category is 6.58.

In [201]:
# Q 1.11 The budget column contains the film's budget. What is the median budget for all the films listed? Give the answer as an integer.
df_movie.budget = df_movie.budget.str.replace('$', '')
df_movie.budget = pd.to_numeric(df_movie.budget)
df_movie.budget
print('The median budget for all the films listed is $', int(df_movie.budget.median()), '.', sep = '')

The median budget for all the films listed is $15000000.


## Task 2. Problem-solving

In [202]:
# Q 2.1 Download the event_data.csv dataset, which contains data on the use of the mobile application of users who registered from July 29 to September 1, 2019:
# Load file event_data.csv into dataframe and inspect.
df_event = pd.read_csv('event_data.csv')
df_event.event_date = pd.to_datetime(df_event.event_date)
print(df_event.dtypes)
print(df_event.shape[0], 'rows')
df_event.head()

user_id                    object
event_date         datetime64[ns]
event_type                 object
purchase_amount           float64
dtype: object
79742 rows


user_id          event_date    event_type  purchase_amount
0  c40e6a 2019-07-29 00:02:15  registration              NaN
1  a2b682 2019-07-29 00:04:46  registration              NaN
2  9ac888 2019-07-29 00:13:22  registration              NaN
3  93ff22 2019-07-29 00:16:47  registration              NaN
4  65ef85 2019-07-29 00:19:23  registration              NaN

Q 2.2 Highlight user cohorts based on the week of registration in the application. The cohort
identifier should be the week ordinal (for example, the week from July 29 to August 4
should have identifier 31).

Answer: The user cohorts are shown in the following results table:

In [203]:
df_event['year'] = df_event.event_date.dt.year
df_event['week'] = df_event.event_date.dt.week
sqldf('select week as user_cohort, count(*) as users from df_event where event_type = "registration" group by week')

user_cohort  users
0           31   1975
1           32   1952
2           33   2045
3           34   1974
4           35   2050

Q 2.3 How many unique users in the cohort with ID 33?

Answer: By inspection of the table of question 2.2, it can be seen that there are 2045 users in the cohort with ID 33.

In [204]:
# Q 2.4 For each event, highlight the indicator lifetime - the weekly lifetime of the cohort. The lifetime indicator is calculated based on the serial number of the week in which the event is committed, relative to the week of registration. For example, an event committed on August 3 by a user from a cohort of registrants at 31 weeks will be committed on the zero week of lifetime, and an event committed by the same user on August 5 will be committed on the first week of lifetime).
df_registration = df_event[df_event.event_type == 'registration']
df_event = sqldf('select e.*, e.week - r.week as lifetime from df_event as e left join df_registration as r on e.user_id = r.user_id')
df_event.event_date = pd.to_datetime(df_event.event_date)
df_event

user_id          event_date    event_type  purchase_amount  year  week  \
0      c40e6a 2019-07-29 00:02:15  registration              NaN  2019    31   
1      a2b682 2019-07-29 00:04:46  registration              NaN  2019    31   
2      9ac888 2019-07-29 00:13:22  registration              NaN  2019    31   
3      93ff22 2019-07-29 00:16:47  registration              NaN  2019    31   
4      65ef85 2019-07-29 00:19:23  registration              NaN  2019    31   
...       ...                 ...           ...              ...   ...   ...   
79737  930c23 2019-09-01 23:57:41  simple_event              NaN  2019    35   
79738  a84999 2019-09-01 23:57:50  simple_event              NaN  2019    35   
79739  175e4d 2019-09-01 23:59:40  simple_event              NaN  2019    35   
79740  1c2210 2019-09-01 23:59:51  simple_event              NaN  2019    35   
79741  d799dc 2019-09-01 23:59:54  simple_event              NaN  2019    35   

       lifetime  
0             0  
1             0  
2             0  
3             0  
4             0  
...         ...  
79737         3  
79738         2  
79739         3  
79740         2  
79741         0  

[79742 rows x 7 columns]

Q 2.5 Build a summary table of changes in the Retention Rate for cohorts depending on lifetime.

Answer: We assume that the definition of k-week 'retention' of a user is a binary variable that equals 1 when the user has an event in lifetime k, 0 otherwise.  
The following summary table displays the retention rate for each pair (cohort, lifetime).

In [205]:
df_lifetime = pd.DataFrame({'lifetime': df_event.lifetime.unique()})
df_user = df_event.loc[df_event.event_type == 'registration', ['user_id', 'week']]
df_user_lifetime = pd.merge(df_user, df_lifetime, how = 'cross')
df_user_lifetime_exists = sqldf('select distinct user_id, lifetime from df_event')
df_user_lifetime_retention = sqldf('select a.user_id, a.week, a.lifetime, case when b.lifetime is null then 0 else 1 end as retention from df_user_lifetime a left join df_user_lifetime_exists b on a.user_id = b.user_id and a.lifetime = b.lifetime')
df_user_lifetime_retention.rename(columns = {'week': 'cohort'}, inplace = True)
df_retention_rate = pd.pivot_table(  data = df_user_lifetime_retention
                                   , index = ['cohort']
                                   , columns = ['lifetime']
                                   , values = ['retention']
                                   , aggfunc = 'mean'
                                  )
df_retention_rate

retention                                       
lifetime         0         1         2         3        4
cohort                                                   
31             1.0  0.927595  0.629367  0.356962  0.15038
32             1.0  0.929303  0.648053  0.361168  0.00000
33             1.0  0.924205  0.661125  0.000000  0.00000
34             1.0  0.929078  0.000000  0.000000  0.00000
35             1.0  0.000000  0.000000  0.000000  0.00000

Q 2.6 What is the 3 week retention rate for a cohort with ID 32? Give the answer in percent, rounded to 2 decimal places, inclusive.

Answer: As can be seen in the summary table of question 2.5, the 3-week retention rate of cohort 32 is 36%.

Q 2.7 Build a summary table of changes in the indicator ARPPU (Average Revenue Per Paying User) for cohorts depending on lifetime.

Answer: The following summary table displays the ARPPU for each pair (cohort, lifetime).

In [206]:
df_event['cohort'] = df_event.week
df_purchase = df_event[df_event.event_type == 'purchase']
df_ARPPU = pd.pivot_table(  data = df_purchase
                          , index = ['cohort']
                          , columns = ['lifetime']
                          , values = ['purchase_amount']
                          , aggfunc = 'mean'
                          , fill_value = 0
                         )
df_ARPPU = round(df_ARPPU, 2)
df_ARPPU

purchase_amount                            
lifetime               0      1      2      3      4
cohort                                              
31                 28.13   0.00   0.00   0.00   0.00
32                 30.22  29.51   0.00   0.00   0.00
33                 30.69  29.77  30.52   0.00   0.00
34                 30.78  30.52  29.93  29.14   0.00
35                 30.72  30.93  29.17  29.53  30.45

Q 2.8 What is the 3-week ARPPU of a cohort with ID 31? Give the answer with a floating point number, rounded to 2 decimal places, inclusive.

Answer: By inspection of the table for question 2.7, the 3-week ARPPU of a cohort with ID 31 is 0.

In [207]:
# Q 2.9 What is the median time between user registration and first purchase? Give the answer in seconds (!) As an integer.
df_purchase1 = df_event[df_event.event_type == 'purchase']
df_purchase1 = sqldf('select user_id, min(event_date) as purchase1_date from df_purchase1 group by user_id')
df_purchase1 = sqldf('select a.user_id, a.purchase1_date, b.event_date from df_purchase1 a inner join df_registration b on a.user_id = b.user_id')
df_purchase1.purchase1_date = pd.to_datetime(df_purchase1.purchase1_date)
df_purchase1.event_date = pd.to_datetime(df_purchase1.event_date)
df_purchase1['time_registration_purchase1'] = (df_purchase1.purchase1_date - df_purchase1.event_date).dt.seconds
median_time = df_purchase1.time_registration_purchase1.median()
print('The median time between user registration and first purchase is', int(median_time), 'seconds.')

The median time between user registration and first purchase is 43623 seconds.


## Task 3. Answering student questions

How would you answer the student's question below? Your task is to get your message across in such a way that a beginner can understand your explanation. You can do this any way you want (pictures, GIFs, metaphors, anything) so long as it makes your explanation clear. Indicate how much time you spent completing this task.

> **What is the difference between DataFrame and Series?**
>

**Answer**

A Python pandas data frame is a table structure with rows and columns, similar to a table in Excel.

Each column is a pandas series, which is essentially a one-dimensional array or vector. Each series can be of any data type (float, string, etc.) but only one; there cannot be more than one data type in a series.

A data frame is a collection of series, since each column is a series.

The following code illustrates this.

First, create a simple data frame:

In [208]:
df = pd.DataFrame({'Name': ['Mary', 'Joe', 'Zoe'], 'Income': [80000, 55000, 120000], 'Schooling': [14, 12, 17]})

To see the entire data frame and its data type, use the following code:

In [209]:
print(df)
type(df)

   Name  Income  Schooling
0  Mary   80000         14
1   Joe   55000         12
2   Zoe  120000         17


pandas.core.frame.DataFrame

To inspect just one column:

In [210]:
print(df.Name)
type(df.Name)

0    Mary
1     Joe
2     Zoe
Name: Name, dtype: object


pandas.core.series.Series

Note how the data type of an individual column is *series*.

## Task 4.

You are given two random variables X and Y.

E(X) = 0.5, Var(X) = 2
E(Y) = 7, Var(Y) = 3.5
cov (X, Y) = -0.8

Find the variance of the random variable Z = 2X - 3Y

#### Solution

Apply the formulae:  
$var(aX) = a^2var(X)$  
$var(X + Y) = var(X) + var(Y) + 2cov(X,Y)$  
$cov(aX, Y) = acov(X, Y)$

$$var(Z) = var(2X) + var(-3Y) + 2cov(2X, -3Y)$$
$$\therefore var(Z) = 2^2var(X) + (-3)^2var(Y) + 2 \cdot 2 \cdot (-3) cov(X, Y)$$

Substitute the given values:

$$var(Z) = 4 \cdot 2 + 9 \cdot 3.5 + 12 \cdot (-0.8)$$

$$\therefore var(Z) = 29.9$$


## Task 5.
It is reasonable to assume that the "expected value" of the dependent variable, y, is the real value available in the testing sample. Therefore, the differences in predictions, $\hat y_i$, relative to the expected values, $y_i$, are:
$$\hat y_i - y_i = 0.5, \ for \ 400 \ observations \ ... \ 5.1$$
$$\hat y_i - y_i = -0.7, \ for \ 100 \ observations \ ... \ 5.2$$

By definition, the MSE is given by:
$$MSE = \frac{1}{n}\sum_{i=1}^n(y_i - \hat y_i)^2 \ ... \ 5.3$$

This sum can be divided into the sum of 400 observations for which $\hat y_i - y_i = 0.5$ plus the sum of 100 observations for which $\hat y_i - y_i = -0.7$. Therefore, the MSE for Omer's model is given by:

$$MSE_O = \frac{1}{500} \left[\sum_{i=1}^{400}(y_i - \hat y_i)^2 + \sum_{i=401}^{500}(y_i - \hat y_i)^2 \right]$$

$$\therefore MSE_O = \frac{1}{500} \left[\sum_{i=1}^{400}(0.5)^2 + \sum_{i=401}^{500}(-0.7)^2 \right]$$

$$\therefore MSE_O = \frac{1}{500} \left[400 \cdot 0.25 + 100 \cdot 0.49 \right]$$

$$\therefore MSE_O = 0.298$$

Limor's improved regression model is given by:

$$z_i = \hat y_i + c \ ... \ 5.4$$

where c is the added constant.

Limor wishes to find the value of c that minimises her MSE. Therefore, Limor expresses her MSE, $MSE_L$, in terms of c by using her corrected regression model:

$$MSE_L = \frac{1}{n}\sum_{i=1}^n(y_i - z_i)^2$$

$$\therefore MSE_L = \frac{1}{n}\sum_{i=1}^n(y_i - \hat y_i - c)^2$$

$$\therefore MSE_L = \frac{1}{500} \left[\sum_{i=1}^{400}(y_i - \hat y_i - c)^2 + \sum_{i=401}^{500}(y_i - \hat y_i - c)^2 \right]$$

$$\therefore MSE_L = \frac{1}{500} \left[\sum_{i=1}^{400}(0.5 - c)^2 + \sum_{i=401}^{500}(-0.7 - c)^2 \right]$$

$$\therefore MSE_L = \frac{1}{500} \left[400 \cdot (0.25 - c + c^2) + 100 \cdot (0.49 + 1.4c + c^2) \right]$$

$$\therefore MSE_L = \frac{1}{500} \left[(100 - 400c + 400c^2) + (49 + 140c + 100c^2) \right]$$

$$\therefore MSE_L = \frac{1}{500} \left[500c^2 - 260c + 149 \right]$$

$$\therefore MSE_L = c^2 - 0.52c + 0.298$$

Express $MSE_L$ as a function of c:
    
$$MSE_L(c) = c^2 - 0.52c + 0.298 \ ... \ 5.5$$

Use differential calculus to find the value of c that minmimises c:
    
$$MSE_L'(c) = 2c - 0.52 = 0$$

$$\therefore c = 0.26$$

To check that this is a local minimum, compute the second derivative at c = 0.26:

$$MSE_L''(c) = 2 > 0$$

The second derivative is greater than zero; therefore c = 0.26 is a local **minimum**.

From equation 5.5, Limor's MSE for c = 0.26 is:

$$MSE_L(0.26) = 0.26^2 - 0.52 \cdot 0.26 + 0.298$$

$$\therefore MSE_L(0.26) = 0.2304$$

To conclude, the values of Omer's and Limor's MSE are respectively:

$$MSE_O = 0.298$$
$$MSE_L = 0.2304$$

Clearly, Limor obtained a lower MSE.


*** End ***